### PR TITLE
Honor attributes on fields

### PR DESCRIPTION
**Pull request by jcouv**
Part of https://github.com/dotnet/roslyn/issues/35816 (work items for attributes)

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs [hunk 1]
```diff
@@ -20,6 +20,59 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols.Metadata.PE
...
+        {
+            // Layout:
+            // |..............................|vvvvv|
+            //
+            // f = FlowAnalysisAnnotations. 5 bits (4 value bits + 1 completion bit).
+
+            private const int HasDisallowNullAttribute = 0x1 << 0;
+            private const int HasAllowNullAttribute = 0x1 << 1;
+            private const int HasMaybeNullAttribute = 0x1 << 2;
+            private const int HasNotNullAttribute = 0x1 << 3;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs]: Handling individual bits here is not necessary since `FlowAnalysisAnnotations` is already stored as a set of bits (see `PEParameterSymbol.PackedFlags`).

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs [hunk 2]
```diff
@@ -20,6 +20,59 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols.Metadata.PE
...
+            private const int FlowAnalysisAnnotationsCompletionBit = 0x1 << 4;
+
+            private int _bits;
+
+#if DEBUG
+            static PackedFlags()
+            {
+                // Verify a few things about the values we combine into flags.  This way, if they ever
+                // change, this will get hit and you will know you have to update this type as well.
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs]: Consider simplifying or eliminating this comment. In [https://github.com/dotnet/roslyn/pull/36152](https://github.com/dotnet/roslyn/pull/36152), all similar comments have been changed to "Verify masks are sufficient for values."

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs]: PEParameterSymbol can store all the possible annotations, but here we only need to store only four, which are not even contiguous (see below). So I think we have to deal with individual bits/flags.
```C#
    [Flags]
    internal enum FlowAnalysisAnnotations
    {
        None = 0,
        AllowNull = 1 << 0,
        DisallowNull = 1 << 1,
        MaybeNullWhenTrue = 1 << 2,
        MaybeNullWhenFalse = 1 << 3,
        MaybeNull = MaybeNullWhenTrue | MaybeNullWhenFalse,
        NotNullWhenTrue = 1 << 4,
        NotNullWhenFalse = 1 << 5,
        NotNull = NotNullWhenTrue | NotNullWhenFalse,
        AssertsTrue = 1 << 6,
        AssertsFalse = 1 << 7,
    }
```

---
In reply to: [296767023](https://github.com/dotnet/roslyn/pull/36704#discussion_r296767023) [](ancestors = 296767023)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 3]
```diff
@@ -26346,6 +26346,50 @@ void symbolValidator(ModuleSymbol m)
...
+    [NotNull]public static string? notNull = null;
+}
+";
+
+            var lib = CreateNullableCompilation(new[] { DisallowNullAttributeDefinition, AllowNullAttributeDefinition, MaybeNullAttributeDefinition, NotNullAttributeDefinition, lib_cs });
+
+            var source = @"
+class D
+{
+    static void M1<T>(T t1)
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: `T` is not used.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 4]
```diff
@@ -26020,6 +26224,199 @@ public class C
...
+    }
+    static void M5<T>(T? t5) where T : struct
+    {
+        new COpen<T?>().f1 = t5; // 8
+        new CStruct<T>().f5 = t5; // 9
+        if (t5 == null) return;
+        new COpen<T?>().f1 = t5;
+        new CStruct<T>().f5 = t5;
+    }
+}";
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Consider testing:
```
static void M6<T>([MaybeNull]T t6) where T : class
{
    new CClass<T>().f2 = t6;
}
static void M7<T>([NotNull]T? t7) where T : class
{
    new CClass<T>().f2 = t7;
}
...
```

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs [hunk 5]
```diff
@@ -20,6 +20,56 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols.Metadata.PE
...
+            private const int HasMaybeNullAttribute = 0x1 << 2;
+            private const int HasNotNullAttribute = 0x1 << 3;
+            private const int FlowAnalysisAnnotationsCompletionBit = 0x1 << 4;
+
+            private int _bits;
+
+#if DEBUG
+            static PackedFlags()
+            {
+                Debug.Assert(EnumUtilities.ContainsAllValues<FlowAnalysisAnnotations>(0xFF));
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs]: The assert seems unrelated to the number of bits stored.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs]: I'd kept the assertion to alert us if the enum changes, so that we consider adding handling here. That said, that seems superfluous, so I'll remove.

---
In reply to: [296972213](https://github.com/dotnet/roslyn/pull/36704#discussion_r296972213) [](ancestors = 296972213)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 6]
```diff
@@ -19389,6 +19389,95 @@ public class CClass<TClass> where TClass : class
...
+            // Warn on misused nullability attributes (f2, f3, f4, f5)? https://github.com/dotnet/roslyn/issues/36073
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public class COpen<TOpen>
+{
+    [MaybeNull]public TOpen f1 = default!;
+}
+public class CClass<TClass> where TClass : class
+{
+    [MaybeNull]public TClass f2 = null!;
```
Comment (agocke) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Out of scope, but I wonder how common it will be for people to misuse the attributes like this.

I think any usage of [MaybeNull] on a non-annotated reference type is probably wrong. The correct attribute usage would probably be to annotate as nullable, right?

File: src/Compilers/CSharp/Portable/Symbols/FieldSymbol.cs [hunk 7]
```diff
@@ -56,6 +56,8 @@ public TypeWithAnnotations TypeWithAnnotations
             }
         }
 
+        public abstract FlowAnalysisAnnotations FlowAnalysisAnnotations { get; }
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Symbols/FieldSymbol.cs]: This name was pretty confusing for me at first. It doesn't really say anything about nullable or that it's only about the attributes, not `?`

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 6 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: In this specific case, it seems okay to use the [MaybeNull]. That indicates that you could get a `null` out, but would be disallowed to put a `null` in.
I don't think that's very useful, but it is not strictly equivalent to saying `TClass? f2 = null;`

File: src/Compilers/CSharp/Portable/Symbols/FieldSymbol.cs (hunk 7 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/FieldSymbol.cs]: This only represents the annotations from user-specified attributes. I agree this name is a bit confusing (we have annotations like `?` and flow analysis annotations like `[DisallowNull]`). I'll think about this for a subsequent PR.
Thanks

**Comment by jcouv** [comment]
I'll hold this PR until either https://github.com/dotnet/roslyn/pull/36784 (fix for metadata loading on simple value types) or https://github.com/dotnet/roslyn/pull/36787 (reverting third metadata optimization) is in.
